warehouse/legacy/api/json.py

```diff
@@ -205,3 +207,82 @@ def json_release_slash(release, request):
         ),
         headers=_CORS_HEADERS,
     )
+
+
+@view_config(
+    route_name="legacy.api.json.token.new",
+    renderer="json",
+    require_methods=["POST"],
+    uses_session=True,
+    permission="manage:user",
+    require_csrf=False,
+)
+def create_token(request):
+    def _fail(message):
+        request.response.status_code = 400
+        return {"success": False, "message": message}
+
+    if request.authenticated_userid is None:
+        return _fail("invalid authentication token")
+
+    # Sanity-check our JSON payload. Ideally this would be done in a form,
+    # but WTForms isn't well equipped to handle JSON bodies.
+    try:
+        payload = request.json_body
+        if not isinstance(payload, dict):
+            raise ValueError
+    except Exception:
+        return _fail("invalid payload")
+
+    macaroon_service = request.find_service(IMacaroonService, context=None)
+
+    form = CreateMacaroonForm(
+        **payload,
```

Comment:
Passing `**` from attacker provided JSON is a risky move. It lets the attacker control which args are provided. If `CreateMacaroonForm` adds a new optional argument this can turn into a vulnerability. It ought to be passed as `data=payload` or explicitly unpacked.

Reply:
Yep, excellent point. This PR needs a decent amount of reworking (and discussion, w/r/t to the semantics of "one time use" for uploads of multiple files per release).